### PR TITLE
ARROW-13995: [R] Bindings for join node

### PR DESCRIPTION
**Pull request by nealrichardson**
This is based on #11150. 

Among the issues observed:

* Dictionary columns aren't allowed even in the left data, though you can first `Project` to remove them
* Duplicate column names aren't allowed at all, even though there is a provision for deduping with a prefix

### CODE REVIEW

File: r/R/dplyr-join.R [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                                        keep = FALSE) {
+  do_join(x, y, by, copy, suffix, ..., keep = keep, join_type = "LEFT_ANTI")
+}
+anti_join.Dataset <- anti_join.ArrowTabular <- anti_join.arrow_dplyr_query
+
+handle_join_by <- function(by, x, y) {
+  if (is.null(by)) {
+    return(set_names(intersect(names(x), names(y))))
+  }
+  stopifnot(is.character(by))
```
Comment (jonkeane) [inline r/R/dplyr-join.R:116]: This is not a comment on this code, but it never occurred to me that it's a bit funny that `by` isn't tidy-evaled and must be character strings (in dplyr!)

Comment (jonkeane) [reply r/R/dplyr-join.R:116]: It makes sense for the `by = c(col_x = "col_y")` interface, but a bit(?) odd for `by = "col"` case

File: r/tests/testthat/test-dplyr-join.R [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+skip_if_not_available("dataset")
+
+library(dplyr)
+
+left <- example_data
+# Error: Invalid: Dictionary type support for join output field is not yet implemented, output field reference: FieldRef.Name(fct) on left side of the join
```
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-join.R:23]:
```suggestion
# Error: Invalid: Dictionary type support for join output field
# is not yet implemented, output field reference: FieldRef.Name(fct)
# on left side of the join
```

File: r/tests/testthat/test-dplyr-join.R [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    left
+  )
+  expect_dplyr_equal(
+    input %>%
+      left_join(
+        to_join %>%
+          rename(the_grouping = some_grouping),
+        by = c(some_grouping = "the_grouping")
+      ) %>%
+      collect(),
```
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-join.R:57]:
```suggestion
    input %>%
      rename(the_grouping = some_grouping) %>%
      left_join(
        to_join
        by = c(the_grouping = "some_grouping")
      ) %>%
      collect(),
```

This reads slightly easier to me, unless you were also intending to test the execution order of having a pipeline as the `y` argument of the join?

File: r/tests/testthat/test-dplyr-join.R [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+})
+
+test_that("anti_join", {
+  expect_dplyr_equal(
+    input %>%
+      anti_join(to_join) %>%
+      collect(),
+    left
+  )
+})
```
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-join.R:108]: I can do this when I rebase, but should we have tests for the error handling as well (e.g. columns named but not in `x` or `y`, etc.)?

**Review by jonkeane — COMMENTED** [review]
Looks good, a few comments / suggestions. Once #11150 merges I'll rebase + push

File: r/tests/testthat/test-dplyr-join.R (hunk 4 above)
Comment (nealrichardson) [reply r/tests/testthat/test-dplyr-join.R:108]: Oh yeah, there are lots of missing tests, and these datasets aren't ideal for testing joins. We may want to just use the datasets that dplyr uses in its tests or something.

**Pull request closed without merging**